rskj-core/src/main/java/co/rsk/net/AsyncNodeBlockProcessor.java

```diff
@@ -0,0 +1,206 @@
+import co.rsk.crypto.Keccak256;
+import co.rsk.net.sync.SyncConfiguration;
+import co.rsk.validators.BlockValidator;
+import org.ethereum.core.Block;
+import org.ethereum.core.Blockchain;
+import org.ethereum.core.ImportResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * AsyncNodeBlockProcessor processes blockchain blocks that are received from other nodes.
+ * If a block passes validation, it will immediately be propagated to other nodes and its execution will be scheduled on a separate thread.
+ * Blocks are being executed and connected to the blockchain sequentially one after another.
+ * <p>
+ * If a block is not ready to be added to the blockchain, it will be on hold in a BlockStore.
+ * <p>
+ */
+public class AsyncNodeBlockProcessor extends NodeBlockProcessor implements InternalService, Runnable {
+
+    private static final Logger logger = LoggerFactory.getLogger("asyncblockprocessor");
+
+    private final BlockingQueue<BlockInfo> blocksToProcess = new LinkedBlockingQueue<>();
+
+    private final BlockValidator blockRelayValidator;
+
+    private final Thread thread = new Thread(this,"async block processor");
+
+    private final Listener listener;
+
+    private volatile boolean stopped;
+
+    public AsyncNodeBlockProcessor(@Nonnull NetBlockStore store, @Nonnull Blockchain blockchain, @Nonnull BlockNodeInformation nodeInformation,
+                                   @Nonnull BlockSyncService blockSyncService, @Nonnull SyncConfiguration syncConfiguration,
+                                   @Nonnull BlockValidator blockRelayValidator, @Nullable Listener listener) {
+        super(store, blockchain, nodeInformation, blockSyncService, syncConfiguration);
+        this.listener = listener;
+        this.blockRelayValidator = blockRelayValidator;
+    }
+
+    public AsyncNodeBlockProcessor(@Nonnull NetBlockStore store, @Nonnull Blockchain blockchain, @Nonnull BlockNodeInformation nodeInformation,
+                                   @Nonnull BlockSyncService blockSyncService, @Nonnull SyncConfiguration syncConfiguration,
+                                   @Nonnull BlockValidator blockRelayValidator) {
+        this(store, blockchain, nodeInformation, blockSyncService, syncConfiguration, blockRelayValidator, null);
+    }
+
+    @Override
+    public BlockProcessResult processBlock(@Nullable Peer sender, @Nonnull Block block) {
+        final Instant start = Instant.now();
+
+        final long blockNumber = block.getNumber();
+        final String blockHash = block.getPrintableHash();
+        final String peer = sender != null ? sender.getPeerNodeID().toString() : "N/A";
+
+        if (store.hasBlock(block)) {
```

Comment:
The PoW rule should be validated before this. 

It would be easy to DoS the nodes in the current state: RSKIP92 allows the creation of a block with the same hash but with invalid PoW, and the current code would allow it to reach the store actually forcing the valid block to be rejected.

Reply:
@donequis, I've separated block validation onto two stages. Block header validation now happens before `store.hasBlock(block)`, and it includes PoW and block timestamp validation rules. Validation of the block itself and its parent happens later on (as it was before). Let me know about your thoughts.